Add view safeguarding information permission for provider users

steventux opened the pull request:
## Context

As a precursor to delivering permissions at a provider organisation level we need a user level permission which can be used in conjunction with the org level permission.

<!-- Why are you making this change? What might surprise someone about it? -->

## Changes proposed in this pull request

- Adds `view_safeguarding_information` permission to `ProviderPermissions`
- Adds permission checkbox on existing provider user provider permissions form.
- Adds permission checkbox on existing support interface form for managing provider users.
- Restricts visibility of safeguarding information based on the user-level permission for the relevant provider.

<!-- If there are UI changes, please include Before and After screenshots. -->

![image](https://user-images.githubusercontent.com/93511/81805508-e1e43200-9512-11ea-9316-59d9dc1b8c40.png)


## Guidance to review

<!-- How could someone else check this work? Which parts do you want more feedback on? -->

## Link to Trello card

https://trello.com/c/bwEGeJa3/2093-build-access-mgmt-safeguarding
<!-- http://trello.com/123-example-card -->

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

mnacos commented on app/views/provider_interface/provider_users/edit_providers.html.erb:
```diff
@@ -20,6 +20,8 @@
                   <%= ppf.hidden_field :provider_id %>
                   <%= ppf.govuk_check_boxes_fieldset :permissions, legend: { text: 'Choose permissions', size: 's' } do %>
                     <%= ppf.govuk_check_box :manage_users, true, multiple: false, label: { text: 'Manage users' } %>
+                    <%= ppf.govuk_check_box :view_safeguarding_information, true, multiple: false,
```
What does `multiple:` mean here?

steventux replied:
https://govuk-form-builder.netlify.app/form-elements/checkboxes/#generating-a-single-checkbox
I think because this isn't a multiple selection on the same attribute it's necessary.

mnacos commented on spec/models/provider_user_spec.rb:
```diff
@@ -78,6 +78,25 @@
+    let(:provider_user) { create :provider_user, :with_provider }
+
+    it 'is false without the correct permission' do
+      expect(provider_user.can_view_safeguarding_information_for?(build_stubbed(:provider, id: 1))).to be false
```
Would this be slightly better with:

```
      provider = provider_user.providers.first
      expect(provider_user.can_view_safeguarding_information_for?(provider)).to be false
```

Just to prove provider affiliation alone does not grant view rights.

mnacos commented on app/views/provider_interface/application_choices/show.html.erb:
```diff
@@ -52,7 +52,9 @@
 
     <% if FeatureFlag.active?('provider_view_safeguarding') %>
-      <%= render ProviderInterface::SafeguardingDeclarationComponent.new(application_form: @application_choice.application_form)%>
+      <% if current_provider_user.can_view_safeguarding_information_for?(@application_choice.provider) %>
```
I wonder if this should be:

```suggestion
      <% if current_provider_user.can_view_safeguarding_information_for?(@application_choice.offered_option.provider) %>
```

offered_option will return the original course option unless an alternative one has been offered by the provider. The alternative one may be belong to a different provider that the user is associated with.

duncanjbrown replied:
This is a good question! I think we should tackle it separately to this PR, though?

mnacos commented on app/models/support_interface/provider_user_form.rb:
```diff
@@ -66,7 +66,11 @@ def provider_permissions=(attributes)
-        permission.manage_users = form.provider_permission.fetch(:manage_users, false)
+
+        ProviderPermissions::VALID_PERMISSIONS.each do |permission_name|
+          permission.send("#{permission_name}=", form.provider_permission.fetch(permission_name, false))
```
I couldn't decide at first, but I think I like this. :)

steventux replied:
Metaprogramming considered harmful! 
I wasn't sure about it either, but this seems like a valid use as it simplifies adding further permissions using the `ProviderPermissions::VALID_PERMISSIONS` constant.

duncanjbrown commented on spec/system/provider_interface/manage_provider_user_permissions_spec.rb:
```diff
@@ -90,13 +95,31 @@ def when_i_remove_a_permissions_from_a_provider_user
+      expect(page).to have_content 'View safeguarding information'
+    end
+
+    expect(@managed_user.provider_permissions.first.view_safeguarding_information).to be true
```
It would be good not to need to inspect the DB as part of a system spec. Could we assert that the checkbox is checked?

steventux replied:
I agree, I think this was a pattern I adopted to debug the system specs quickly but we should use expect the UI to reflect db state. Line 117 does this as it's the output from the permissions list component.

duncanjbrown commented on app/components/provider_interface/provider_permissions_list_component.html.erb:
```diff
@@ -1,14 +1,18 @@
-      </li>
+    <%
+      humanized_permissions = [].tap do |ary|
+        ProviderPermissions::VALID_PERMISSIONS.each do |permission_name|
```
This is quite gnarly for the view — maybe a helper method on the component?

duncanjbrown commented on app/components/provider_interface/provider_permissions_list_component.html.erb:
```diff
@@ -1,14 +1,18 @@
 <% visible_provider_permissions.each do |provider_permission| %>
```
I just noticed this hides the provider altogether when no permissions are set:

When manage users permission is set for Oriel:
![Screenshot 2020-05-18 at 09 24 44](https://user-images.githubusercontent.com/642279/82190797-7e2e8000-98e9-11ea-98c4-a362551830d8.png)

When no permissions are set:
![Screenshot 2020-05-18 at 09 25 53](https://user-images.githubusercontent.com/642279/82190869-96060400-98e9-11ea-984e-755bf3a8decd.png)

I'll add a separate card

steventux replied:
That's odd I definitely see the No permissions message:
![image](https://user-images.githubusercontent.com/93511/82209705-a035fb80-9905-11ea-9bbd-84505a2b785d.png)

duncanjbrown replied:
Probably me mixing up my branches locally, I'll check

duncanjbrown replied:
My bad: I removed `manage users` permission for my current user on one of the provider I had access to. Not a bug

duncanjbrown commented on spec/system/support_interface/managing_provider_users_spec.rb:
```diff
@@ -179,6 +187,14 @@ def and_they_should_be_able_to_manage_users
   end
 
+  def and_they_should_be_able_to_view_safeguarding_information
+    expect(@user.reload.provider_permissions.view_safeguarding_information.first.provider).to eq(@provider)
```
Could we get away without this assertion?

duncanjbrown commented on spec/models/provider_user_spec.rb:
```diff
@@ -78,6 +78,25 @@
+
+    it 'is false without the correct permission for the given provider' do
+      provider_user.provider_permissions.update_all(view_safeguarding_information: true)
+      expect(provider_user.can_view_safeguarding_information_for?(build_stubbed(:provider, id: 1))).to be false
```
Do we need an `id` on this one?

steventux commented:
@mnacos @duncanjbrown I think I've addressed all comments apart from https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/2057#discussion_r425233371 which may need some discussion.

tijmenb commented on spec/system/provider_interface/see_individual_application_spec.rb:
```diff
@@ -44,6 +45,11 @@ def and_the_safeguarding_declaration_feature_flag_is_active
+    ProviderUser.find_by(dfe_sign_in_uid: 'DFE_SIGN_IN_UID')
+      .provider_permissions.update_all(view_safeguarding_information: true)
+  end
+
```
@steventux I just accidentally broke this feature while rebasing, but the tests still pass. Can you add a test that the provider user _can't_ see the safeguarding data without the correct permission?

steventux replied:
https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/2097
